referral_module/api/serializers.py

```diff
@@ -0,0 +1,85 @@
+from rest_framework import serializers
+from referral_module import models
+from django.contrib.auth.models import User
+
+
+class UserSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = User
+        fields = ('id', 'username', )
+
+
+class UserReferrerSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = models.UserReferrer
+        fields = ('user', 'key', )
+
+
+class CampaignSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = models.Campaign
+        fields = ('id', 'key', 'description', 'is_active', 'created',)
+
+
+class CampaignDetailsSerializer(serializers.ModelSerializer):
+
+    user_referrers = serializers.SerializerMethodField()
+
+    class Meta:
+        model = models.Campaign
+        fields = (
+            'id', 'key', 'user_referrers',
+        )
+
+    def get_user_referrers(self, obj):
+        users = User.objects.all()
+        result = []
+        for user_referrer in self.instance.user_referrers.values():
+            username = users.get(id=user_referrer['user_id'])
+            referrer_key = user_referrer['key']
+            result.append(
+                {
+                   'user': username.username,
+                   'referrer_key': referrer_key,
+                }
+            )
+        return result
+
+
+class CampaignUserReferralSerializer(serializers.ModelSerializer):
+
+    user_referrer = serializers.SerializerMethodField(read_only=True)
+
+    class Meta:
+        model = models.Campaign
+        fields = (
+            'user_referrer',
+        )
+
+    def get_user_referrer(self, obj):
+        request = self.context.get("request")
+
+        if request and hasattr(request, "user"):
+            user = request.user
+        else:
+            raise Exception('Authentication required')
+        campaign = obj
+        if user and obj:
+            user_referrers_qs = models.UserReferrer.objects.filter(
+                user=user,
+                campaign=campaign
+            )
+        else:
+            return ''
+        return user_referrers_qs.first().key if user_referrers_qs.exists() else ''
+
+    def generate_user_referrer(self, request, campaign):
```

Comment:
`campaign`  is might be a bit redundant, your serializer should already have an access to the instance of the campaign.